### ISSUE TITLE
The distributed compiler sources do not come with an up-to-date lockfile

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#116229]
After installing the `rustc-src` component, trying to run `cargo metadata` on them will fail if the sources are installed in a read-only location (such as on Nix, or when `cargo metadata` is run in a sandbox). This could be avoided by either fixing cargo (https://github.com/rust-lang/cargo/issues/10096), or by adjusting the rustc-src component to work around this behavior of cargo, by shipping an up-to-date lockfile.

Part of the problem seems to be that the cargo crates are not in a workspace any more when they are inside rustc-src. However, sadly, even if one puts a `Cargo.toml` (with all references to `library/` and `src/tools/` removed) into the right place, Cargo is still not happy with the lockfile -- it insists on removing everything unnecessary from this lockfile. So producing a lockfile that cargo is happy with could be rather tricky.

**Comment by jonhoo** [comment]
This feels somewhat-related to https://github.com/rust-lang/rust/issues/95736

**Comment by RalfJung** [comment]
Yeah this is kind of the rustc-src version of that rust-src issue (note the rustc vs rust).

Somehow the rust-src issue doesn't seem to affect rust-analyzer or it has another way to work on a read-only file system; it's only with rustc-src that I am having trouble.

**Comment by jonhoo** [comment]
Now that you mention it, I seem to remember there's some special handling for rust-src in RA iirc for exactly this reason :thinking:
